src/lab.rs

```diff
@@ -190,33 +198,75 @@ pub fn test_mutants(
     Ok(lab_outcome)
 }
 
-fn test_timeout(baseline_test_duration: Option<Duration>, options: &Options) -> Duration {
-    if let Some(timeout) = options.test_timeout {
+#[derive(Copy, Clone)]
+struct Timeouts {
+    build: Duration,
+    test: Duration,
+}
+
+fn phase_timeout(
+    explicit_timeout: Option<Duration>,
+    check_only_timeout: Duration,
+    baseline_duration: Option<Duration>,
+    minimum: Duration,
+    multiplier: f64,
+    phase_name: &str,
+    options: &Options,
+) -> Duration {
+    if let Some(timeout) = explicit_timeout {
         timeout
     } else if options.check_only {
-        Duration::ZERO
-    } else if let Some(baseline_test_duration) = baseline_test_duration {
-        let timeout = max(
-            options.minimum_test_timeout,
-            Duration::from_secs(
-                (baseline_test_duration.as_secs_f64()
-                    * options.test_timeout_multiplier.unwrap_or(5.0))
-                .round() as u64,
-            ),
+        check_only_timeout
+    } else if let Some(baseline_duration) = baseline_duration {
+        let mut timeout = max(
+            minimum,
+            Duration::from_secs_f64(baseline_duration.as_secs_f64() * multiplier),
         );
+
+        // Round large timeouts to a nice even number, but round smaller timeouts to 2DP
+        timeout = if timeout.as_secs() >= 1 {
+            Duration::from_secs(timeout.as_secs_f64().round() as u64)
+        } else {
+            Duration::from_secs_f64((timeout.as_secs_f64() * 100.0).round() / 100.0)
+        };
+
         if options.show_times {
             info!(
-                "Auto-set test timeout to {}",
+                "Auto-set {phase_name} timeout to {}",
                 humantime::format_duration(timeout)
             );
         }
         timeout
     } else {
-        warn!("An explicit timeout is recommended when using --baseline=skip; using 300 seconds by default");
+        warn!("An explicit {phase_name} timeout is recommended when using --baseline=skip; using 300 seconds by default");
         Duration::from_secs(300)
     }
 }
 
+fn test_timeout(baseline_duration: Option<Duration>, options: &Options) -> Duration {
+    phase_timeout(
+        options.test_timeout,
+        Duration::ZERO,
+        baseline_duration,
+        options.minimum_test_timeout,
+        options.test_timeout_multiplier.unwrap_or(5.0),
+        "test",
+        options,
+    )
+}
+
+fn build_timeout(baseline_duration: Option<Duration>, options: &Options) -> Duration {
+    phase_timeout(
+        options.build_timeout,
+        Duration::MAX,
```

Comment:
This seems to say that with `--check`, i.e. running just `cargo check --tests` on each mutant, the build timeout is unlimited. I haven't checked yet but I suspect a const fn that hangs the compiler might also hang the checks? Maybe we should check that it doesn't.

And, maybe the logic for the check timeout should be the same as for non-check builds, i.e. a multiplier of the baseline?

Reply:
My intent was for the `check` timeouts to always be the same as the `build`. It's definitely true that the `check` hangs when a const function is divergent.

I didn't really test that the `check` works like this, though. Might be an area to improve the testing in.

Reply:
This code here was a little backwards, the code change here
https://github.com/sourcefrog/cargo-mutants/pull/345/files/9cf221652a74ac8dff8fbe95a01b5909a56b3e0d#diff-43aa7803b24379ea66e0395da5f5baf8ea6eef4bd298b984bfa7c992f0de312dR309
made sure that the timeout used for `check` phases is whatever the final `build` timeout is.

I'll delete the special handing here, and add some tests to assert how it works.